### ISSUE TITLE
No components, pipes, etc. found from within an Angular CLI library

### ISSUE THREAD

**Issue by arobinson** [angular/vscode-ng-language-service#340]
I just refactored our code to use an Angular library. At runtime, everything is working and rendered correctly, but this VSCode extension is saying that none of the components, directives, pipes, etc. can be found when the template HTML is open.

Example error:
```
nw-messages' is not a known element:
1. If 'nw-messages' is an Angular component, then verify that it is part of this module.
2. If 'nw-messages' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@NgModule.schemas' of this component to suppress this message.
```

And:
```
The pipe 'NwTranslationPipe' could not be found Angular
```

Our project layout is an Angular 7 (CLI) application with an Angular Library child project. None of the items from the Angular library are being found by this extension. Components from the application are found, it is only the library ones that are not working.

Once I get a chance I'll create a project on (like github) to reproduce the issue

Are there any known issues with this extension and using an Angular library in a child project of an angular application?

**Comment by arobinson** [comment]
Application to reproduce the issue:
https://github.com/arobinson/vscodeNgLanguageService340

Follow the steps in the readme to see the issue

**Comment by JohnnyDevNull** [comment]
You have to build it first, then reload VSCode. After this steps it should work, but keep in mind that the build files are locked after it, so hou have to "Reload > Remove Build > Rebuild > Reload" in this cycle... it is a bad workflow i know, but actually the plugin locks the files with the duplicate TSServer running.

**Comment by Urastor** [comment]
Having the same problem when using RouterModule. I have an app-module with an app-routing-module and import a feature-module with a feature-routing-module. In my feature-component template, I can't find any references to any directives or pipes. I even get an error. Running the app works fine.

![Anmerkung 2019-10-11 111055](https://user-images.githubusercontent.com/7266207/66641319-c7637680-ec1a-11e9-9b28-56b9ae5ae9b8.png)

**Comment by kyliau** [comment]
Possible root cause: https://github.com/angular/vscode-ng-language-service/issues/457

**Comment by granttw** [comment]
I am having a similar experience with a slight difference.

I created a project very similar to @arobinson.  A workspace with a library project and an application project.  I take advantage of the tsconfig.json path maps and import the library module without publishing it to npm.  If I build the library before adding the import to the application, everything works fine.  If I add a new component to the library, build, and try to use it in the application, I get the error that says it does not exists.

If I delete the dist folder for the library build and rebuild, the error goes away.   This is what I believe @JohnnyDevNull was saying.  

Is this expected behavior?  Am I doing something incorrectly?  If this is expected behavior, it seems like this makes the --watch option on the ng build command for the library useless.

@kyliau  - can you please confirm if this is expected behavior?  Thanks.

**Comment by kyliau** [comment]
@granttw Based on your description, it looks like the language service is not picking up new metadata.json from the library after a new component is added.
The reason is because we are only watching for changes in `d.ts` file. When a new component is added and compiled, the new `.d.ts` file gets picked up, but not the corresponding `metadata.json`.

I can think of two possible solutions:
1. Reload `metadata.json` when the `.d.ts` file changes
2. Wait for Ivy integration, at which point `metadata.json` is no longer needed.

I'll explore (1) for now and see how much work it takes to fix this.

**Comment by JohnnyDevNull** [comment]
@granttw you do everything right, that's the actual workflow with VSCode, only WebStorm does it a bit better, but also there I have to reload in some edge cases.

**Comment by kuncevic** [comment]
Same here. I have a monorepo setup with angular material module living in my shared library. 
So I've got that weird error withing in templates templates on each material component.

![image](https://user-images.githubusercontent.com/1283734/73036046-4e70ed80-3e9e-11ea-91da-7e8332ce2589.png)

**Comment by alberto-chiesa** [comment]
This is impacting our workflow pretty severely. It's strange, but it used to work before 0.900. Does someone knows of any workarounds to get the library discovery work again?
Having lots of components and directives in a library means that a lot of code is reported as errors, but the compilation (both via serve and build) works just fine.

**Comment by kyliau** [comment]
@alberto-chiesa Did you compile your library with [flatModuleOutFile](https://angular.io/guide/angular-compiler-options#flatmoduleoutfile)? Could you please check or paste your `angular.json`?

**Comment by alberto-chiesa** [comment]
@kyliau nope. Should I? Sure, I can copy and paste the `angular.json`, but it's pretty large(ish): 80KB.
But AFAICT, that option shouldn't go in `tsconfig.lib.json`?

**Comment by alberto-chiesa** [comment]
Btw, here it is my tsconfig.lib.json for the referenced library:
[tsconfig.lib.json.txt](https://github.com/angular/vscode-ng-language-service/files/4400998/tsconfig.lib.json.txt)

And the base tsconfig in the root:
[tsconfig.json.txt](https://github.com/angular/vscode-ng-language-service/files/4401002/tsconfig.json.txt)

**Comment by mtgibbs** [comment]
I actually have a similar problem with a `pipe` in from a shared library when adding it to my `shared.module.ts` in the project I'm trying to use it in (Pretty much what was described in #98 ).  While it compiles and runs perfectly fine, but VS Code complains about it.

I grabbed a screenshot of a simple component with my pipe showing an error.  I can probably make up an example project that shows this.  This is with compiling my project with your suggested `flatModuleOutFile` option in `tsconfig`.

![Screen Shot 2020-04-01 at 12 57 30 PM](https://user-images.githubusercontent.com/3111267/78165114-d502f780-7418-11ea-87bc-739fce21b1c9.png)

**Comment by helzgate** [comment]
@alberto-chiesa You can uninstall the extension for now, this is what I did.

**Comment by JohnnyDevNull** [comment]
Maybe this could help you out:

```    "paths": {
      "@your/library": [
        "dist/jp-blog",
        "projects/jp-blog/src",
      ]
```

Extend your path mapping so that the local source get's collected when there is no build done. This improves the library development a lot, because you don't need to build it first. There is only one additinal tweak left to get this work:

1. put a index.ts next to your public_api.ts and move all re-exports from the public_api.ts to the index.ts
2. re-export all from the index.ts in the public_api.ts

Then your local mapping works like a charm!

**Comment by 18clans** [comment]
I ran into this same issue where my custom pipes was not found.  Take a look at my post https://github.com/angular/vscode-ng-language-service/issues/888 to see if it'll resolve yours since I got mine resolved.

**Comment by kyliau** [comment]
This has been fixed by the new Ivy-native language service, released in [v11.1.0](https://github.com/angular/vscode-ng-language-service/releases/tag/v11.1.0).
It's an opt-in feature for now, please give it a try and let us know if you have any feedback.
For the best editor experience, please make sure your project has [`strictTemplates`](https://angular.io/guide/angular-compiler-options#stricttemplates) enabled in `angularCompilerOptions`.
In case you run into similar bug, please file a new issue. I'll close this for now.
